qiskit/pulse/schedule.py

```diff
@@ -1024,13 +1175,90 @@ def instructions(self) -> Tuple[Tuple[int, Instruction]]:
+
+            The sequence of elements is returned in order of addition. Because the first element is
+            schedule first, e.g. FIFO, the returned sequence is roughly time-ordered.
+            However, in the parallel alignment context, especially in
+            the as-late-as-possible scheduling, or :class:`.AlignRight` context,
+            the actual timing of when the instructions are issued is unknown until
+            the :class:`.ScheduleBlock` is scheduled and converted into a :class:`.Schedule`.
+        """
+        blocks = []
+        for elm in self._blocks:
+            if isinstance(elm, Reference):
+                elm = self.references.get(elm.ref_keys, None) or elm
+            blocks.append(elm)
+        return tuple(blocks)
+
+    def _collect_parameters(
+        self,
+        scope: str,
+        add_scope: bool = False,
+    ) -> Dict[Tuple[str, int], Parameter]:
+        """A helper method to return parameters.
+
+        Parameters defined in the nested subroutines are recursively investigated and
+        returned with parameters within the current scope.
+        To distinguish where the parameters are defined, set ``add_scope=True``.
+
+        Args:
+            scope: Name of current scope.
+            add_scope: Set ``True`` to prepend scope to parameter name.
+
+        Returns:
+            All parameters defined under the current scope.
+        """
+        parameters_out = {}
+        for param in self._parameter_manager.parameters:
+            if add_scope:
+                param = _scope_parameter(param, scope, Reference.scope_delimiter)
+            # This avoids unexpected merging of parameter objects defined in different scopes.
+            # Note that the search_parameters method is scope-aware.
+            # However, when the same parameter object is used in multiple scopes,
+            # if it would return one object from a random scope, the search could fail.
+            # This logic evaluates the parameter equality not only with uuid, but also with name.
+            scope_aware_key = param.name, hash(param)
+            parameters_out[scope_aware_key] = param
+
+        for sub_namespace, subroutine in self.references.items():
+            if subroutine is None:
+                continue
+            composite_key = Reference.key_delimiter.join(sub_namespace)
+            full_path = f"{scope}{Reference.scope_delimiter}{composite_key}"
+            sub_parameters = subroutine._collect_parameters(scope=full_path, add_scope=add_scope)
+            parameters_out.update(sub_parameters)
+
+        return parameters_out
 
     @property
-    def parameters(self) -> Set:
-        """Parameters which determine the schedule behavior."""
-        return self._parameter_manager.parameters
+    def parameters(self) -> Tuple[Parameter]:
+        """Return unassigned parameters with raw names."""
+        return tuple(self._collect_parameters(scope="root", add_scope=False).values())
```

Comment:
I think this is a public API change?  Is it absolutely necessary?

Reply:
No. API and behavior doesn't change. It additionally checks subroutines because parameters in there are explicitly unexposed to the current scope.

Reply:
Sorry you are right, this is unnecessary change. Fixed in https://github.com/Qiskit/qiskit-terra/pull/8005/commits/0fc47056ec17df4434ca476fdb2160f08124c7e8